ee/localserver/zta_auth_middleware.go

```diff
@@ -0,0 +1,285 @@
+// The last key is an∂ x25519 key which is the public half of a NaCl box key.
+type chain struct {
+	Links []chainLink `json:"links"`
+	// counterPartyPubEncryptionKey is the last public key in the chain, which is a x25519 key
+	// we set this after we extract and verify in the validate method
+	counterPartyPubEncryptionKey *[32]byte
+}
+
+// chainLink represents a link in a chain of trust
+type chainLink struct {
+	// Payload a b64 url encoded json
+	Payload string `json:"payload"`
+	// Signature is the raw bytes of the signature of the b64 url encoded json
+	Signature string `json:"signature"`
+	// Signed by is the key id "kid" of the key that signed this link
+	SignedBy string `json:"signedBy"`
+}
+
+func (c *chain) UnmarshalJSON(data []byte) error {
+	// incomming data is just an array of chain links
+	var links []chainLink
+	if err := json.Unmarshal(data, &links); err != nil {
+		return fmt.Errorf("failed to unmarshal chain links: %w", err)
+	}
+
+	c.Links = links
+	return nil
+}
+
+// MarshalJSON marshals the chain as a json array of chain links
+func (c *chain) MarshalJSON() ([]byte, error) {
+	// outgoing data is just an aray of chain links
+	bytes, err := json.Marshal(c.Links)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal chain links: %w", err)
+	}
+	return bytes, nil
+}
+
+// validate iterates over all the links in the chain and verifies the signatures and timestamps
+// the root p256 ecdsa key must be in the trusted keys map, the map key is the "kid" of the key
+func (c *chain) validate(trustedKeys map[string]*ecdsa.PublicKey) error {
+	if len(c.Links) == 0 {
+		return errors.New("chain is empty")
+	}
+
+	for i := range len(c.Links) {
+		if c.Links[i].Payload == "" || c.Links[i].Signature == "" {
+			return fmt.Errorf("link %d is missing data or sig", i)
+		}
+	}
+
+	rootKey, ok := trustedKeys[c.Links[0].SignedBy]
+	if !ok {
+		return fmt.Errorf("root key with kid %s not found in trusted keys", c.Links[0].SignedBy)
+	}
+
+	// make a copy of the root key so that we can reassign it
+	parentEcdsa := &ecdsa.PublicKey{
+		Curve: rootKey.Curve,
+		X:     rootKey.X,
+		Y:     rootKey.Y,
+	}
+
+	var lastPayload payload
+
+	for i := range len(c.Links) {
+
+		signature, err := base64.URLEncoding.DecodeString(c.Links[i].Signature)
+		if err != nil {
+			return fmt.Errorf("failed to decode signature: %w", err)
+		}
+
+		if err := echelper.VerifySignature(parentEcdsa, []byte(c.Links[i].Payload), signature); err != nil {
+			return fmt.Errorf("invalid signature: %w", err)
+		}
+
+		payloadBytes, err := base64.URLEncoding.DecodeString(c.Links[i].Payload)
+		if err != nil {
+			return fmt.Errorf("failed to decode payload: %w", err)
+		}
+
+		if err := json.Unmarshal(payloadBytes, &lastPayload); err != nil {
+			return fmt.Errorf("failed to unmarshal payload: %w", err)
+		}
+
+		if lastPayload.ExpirationDate < time.Now().Unix() {
+			return fmt.Errorf("payload %d has expired", i)
+		}
```

Comment:
Do we also need/want to check that `payload.DateTimeSigned` is not too far in the past, or is this time check is sufficient?

Reply:
I believe this is sufficient, the server side is just `payload.DateTimeSigned + TTL`